File: MFAdsDemo_OC/MFAdsDemo_OC/Demo/ExampleController/Template/BaseViewController.m
//
//  BaseViewController.m
//  Example
//
//  Created by CherryKing on 2019/12/20.
//  Copyright © 2019 CherryKing. All rights reserved.
//

#import "BaseViewController.h"
#import "AppDelegate.h"

#define ssRGBHex(rgbValue) [UIColor colorWithRed:((float)((rgbValue & 0xFF0000) >> 16))/255.0 green:((float)((rgbValue & 0xFF00) >> 8))/255.0 blue:((float)(rgbValue & 0xFF))/255.0 alpha:1.0]
#define ssRGBHexAlpha(rgbValue,a) [UIColor colorWithRed:((float)((rgbValue & 0xFF0000) >> 16))/255.0 green:((float)((rgbValue & 0xFF00) >> 8))/255.0 blue:((float)(rgbValue & 0xFF))/255.0 alpha:(a)]


@interface BaseViewController () <UITableViewDataSource, UITableViewDelegate>

@property (nonatomic, strong) UIButton *btnLoad;
@property (nonatomic, strong) UIButton *btnShow;
@property (nonatomic, strong) UIButton *btnLoadAndShow;
@property (nonatomic, strong) UILabel *labNotify;
@property (nonatomic, copy) NSString *text;

@end

@implementation BaseViewController

- (void)viewDidLoad {
    [super viewDidLoad];
    _text = @"";
    self.isOnlyLoad = YES;
    self.view.backgroundColor = [UIColor whiteColor];
        
    if (@available(iOS 15.0, *)) {
        UINavigationBarAppearance *appearance = [[UINavigationBarAppearance alloc] init];
        appearance.backgroundEffect = [UIBlurEffect effectWithStyle:UIBlurEffectStyleRegular];
        self.navigationController.navigationBar.scrollEdgeAppearance = appearance;
        self.navigationController.navigationBar.standardAppearance = appearance;
    }
    
    [self.view addSubview:self.btnLoad];
    [self.view addSubview:self.btnShow];
    [self.view addSubview:self.btnLoadAndShow];
    [self.view addSubview:self.labNotify];
    [self.view addSubview:self.textV];
    [self loadAdWithState:AdState_Normal];

    self.btnLoad.frame = CGRectMake(50, 100, 140, 40);
    self.btnShow.frame = CGRectMake(50, 160, 140, 40);
    self.btnLoadAndShow.frame = CGRectMake(50, 220, 140, 40);
    self.labNotify.frame = CGRectMake(0, 280, self.view.frame.size.width, 40);
    self.textV.frame = CGRectMake(0, 330, self.view.frame.size.width, self.view.frame.size.height - CGRectGetMaxY(self.labNotify.frame) - 20);
}

- (void)setIsOnlyLoad:(BOOL)isOnlyLoad {
    _isOnlyLoad = isOnlyLoad;
    self.btnLoad.hidden = !_isOnlyLoad;
    self.btnShow.hidden = !_isOnlyLoad;
}

- (void)loadAd {
    [self clearText];
    [self loadAdWithState:AdState_Normal];
}

- (void)showAd {}

- (void)loadAndShowAd {
    [self clearText];
    [self loadAdWithState:AdState_Normal];
}

- (void)deallocAd {}

- (void)showProcessWithText:(NSString *)text {
    _text = [_text stringByAppendingString:[NSString stringWithFormat:@"\r\n%@",text]];
    self.textV.text = _text;
}

- (void)showErrorWithDescription:(NSDictionary *)description {
    if (!description) {
        return;
    }
    
    NSString *highlightText = [self getSianKeyWithDic:description];
    _text = [_text stringByAppendingString:[NSString stringWithFormat:@", 失败原因如下:\r\n%@",highlightText]];
    NSMutableAttributedString *attributedStr = [self setupAttributeString:_text highlightText:highlightText];
    _textV.attributedText = attributedStr;
}

- (void)clearText {
    _text = @"";
    self.textV.text = nil;
}

- (void)loadAdWithState:(AdState)state {
    switch (state) {
        case AdState_Normal:
            self.labNotify.text = @"未加载广告";
            break;
        case AdState_Loading:
            self.labNotify.text = @"广告加载中...";
            break;
        case AdState_LoadSucceed:
            self.labNotify.text = @"广告加载成功";
            break;
        case AdState_LoadFailed:
            self.labNotify.text = @"广告加载失败";
            break;

        default:
            break;
    }
}

#pragma mark - pravate
- (NSString *)getSianKeyWithDic:(NSDictionary *)dic {
    //按字典排序
    NSArray* arr = [dic allKeys];
    arr = [arr sortedArrayUsingComparator:^NSComparisonResult(id obj1, id obj2){
        NSComparisonResult result = [obj1 compare:obj2];
        return result == NSOrderedDescending;
    }];
    
    //拼接字符串
    NSMutableArray * strArray =[[NSMutableArray alloc]init];
    [arr enumerateObjectsUsingBlock:^(id  obj, NSUInteger idx, BOOL * _Nonnull stop) {

        NSError *par = dic[obj];
        if (par) {
            NSInteger code = par.code;
            NSString *desc = par.userInfo[NSLocalizedDescriptionKey];
            NSString * desc1 = par.userInfo[@"extra_reason"];
            if (!desc) {
                desc = par.userInfo[@"desc"];
            }
            NSString *errorInfo = [NSString stringWithFormat:@"%@: \ncode: %ld \n错误详情:%@(%@)", obj, code, desc,desc1];
            [strArray addObject:errorInfo];
        }
    }];
    NSString * str = [strArray componentsJoinedByString:@"\r\n"];
    return str;
}

- (NSMutableAttributedString *)setupAttributeString:(NSString *)text highlightText:(NSString *)highlightText {
    
    NSRange hightlightTextRange = [text rangeOfString:highlightText];
    NSMutableAttributedString *attributeStr = [[NSMutableAttributedString alloc] initWithString:text];
    
    if (hightlightTextRange.length > 0) {
        [attributeStr addAttribute:NSForegroundColorAttributeName
                             value:[UIColor redColor]
                             range:hightlightTextRange];
        [attributeStr addAttribute:NSFontAttributeName value:[UIFont boldSystemFontOfSize:15.0f] range:hightlightTextRange];
        
        return attributeStr;
    } else {
        return [[NSMutableAttributedString alloc]initWithString:highlightText];
    }
}

- (UIButton *)btnLoad {
    if (!_btnLoad) {
        _btnLoad = [UIButton buttonWithType:UIButtonTypeSystem];
        _btnLoad.backgroundColor = ssRGBHex(0xf58220);
        _btnLoad.layer.cornerRadius = 5.f;
        _btnLoad.layer.masksToBounds = YES;
        [_btnLoad setTitle:@"load Ad" forState:UIControlStateNormal];
        [_btnLoad setTitleColor:[UIColor whiteColor] forState:UIControlStateNormal];
        [_btnLoad addTarget:self action:@selector(loadAd) forControlEvents:UIControlEventTouchUpInside];
    }
    return _btnLoad;
}

- (UIButton *)btnShow {
    if (!_btnShow) {
        _btnShow = [UIButton buttonWithType:UIButtonTypeSystem];
        _btnShow.backgroundColor = ssRGBHex(0xf58220);
        _btnShow.layer.cornerRadius = 5.f;
        _btnShow.layer.masksToBounds = YES;
        [_btnShow setTitle:@"show Ad" forState:UIControlStateNormal];
        [_btnShow setTitleColor:[UIColor whiteColor] forState:UIControlStateNormal];
        [_btnShow addTarget:self action:@selector(showAd) forControlEvents:UIControlEventTouchUpInside];

    }
    return _btnShow;
}

- (UIButton *)btnLoadAndShow {
    if (!_btnLoadAndShow) {
        _btnLoadAndShow = [UIButton buttonWithType:UIButtonTypeSystem];
        _btnLoadAndShow.backgroundColor = ssRGBHex(0xf58220);
        _btnLoadAndShow.layer.cornerRadius = 5.f;
        _btnLoadAndShow.layer.masksToBounds = YES;
        [_btnLoadAndShow setTitle:@"load and showAd" forState:UIControlStateNormal];
        [_btnLoadAndShow setTitleColor:[UIColor whiteColor] forState:UIControlStateNormal];
        [_btnLoadAndShow addTarget:self action:@selector(loadAndShowAd) forControlEvents:UIControlEventTouchUpInside];
    }
    return _btnLoadAndShow;
}

- (UILabel *)labNotify {
    if (!_labNotify) {
        _labNotify = [UILabel new];
        _labNotify.textColor = ssRGBHex(0xf58220);
        _labNotify.font = [UIFont systemFontOfSize:15];
        _labNotify.textAlignment = NSTextAlignmentCenter;
    }
    return _labNotify;
}

- (UITextView *)textV {
    if (!_textV) {
        _textV = [UITextView new];
        _textV.font = [UIFont systemFontOfSize:15];
        _textV.contentInset = UIEdgeInsetsMake(10, 10, 10, 10);
        _textV.textAlignment = NSTextAlignmentLeft;
        _textV.editable = NO;
    }
    return _textV;
}

@end

File: MFAdsDemo_Swift/MFAdsDemo_Swift/Demo/ExampleController/Template/BaseViewController.swift
//
//  BaseViewController.swift
//  MFAdsDemo_Swift
//
//  Created by cc on 2022/5/31.
//

import UIKit

class BaseViewController : UIViewController {
    
    enum AdState : Int {
        case _Normal // 未加载
        case _Loading // 加载中
        case _LoadSucceed // 加载成功
        case _LoadFailed // 加载失败
    }
    
    public var dic: NSDictionary = [:]
    // 是否正在加载
    public var isLoaded: Bool = false
    
    public lazy var btnLoad: UIButton = {
        let button = UIButton(frame: CGRect(x: 50, y: 100, width: 140, height: 40))
        button.backgroundColor = .blue
        button.setTitle("加载广告", for: .normal)
        button.addTarget(self, action: #selector(handleClickLoadBtn), for: .touchUpInside)
        return button
    }()
    
    public lazy var btnShow: UIButton = {
        let button = UIButton(frame: CGRect(x: 50, y: 160, width: 140, height: 40))
        button.backgroundColor = .blue
        button.setTitle("显示广告", for: .normal)
        button.addTarget(self, action: #selector(handleClickShowBtn), for: .touchUpInside)
        return button
    }()
    
    public lazy var btnLoadAndShow: UIButton = {
        let button = UIButton(frame: CGRect(x: 50, y: 220, width: 140, height: 40))
        button.backgroundColor = .blue
        button.setTitle("加载并显示广告", for: .normal)
        button.addTarget(self, action: #selector(handleClickLoadAndShowBtn), for: .touchUpInside)
        return button
    }()
    
    public lazy var labNotify: UILabel = {
        let lable = UILabel(frame: CGRect(x: 10, y: 280, width: UIScreen.main.bounds.width, height: 40))
        return lable
    }()
    
    public lazy var textV: UITextView = {
        let textView = UITextView(frame: CGRect(x: 0, y: 330, width: self.view.frame.size.width, height: self.view.frame.size.height - self.labNotify.frame.maxY - 20))
        return textView
    }()
    
    public var tempText: String = ""
    // 是否展示某些按钮
    public var isAdLoadAndShow: Bool = false {
      
        didSet {
            if !isAdLoadAndShow {
                btnLoadAndShow.isHidden = !isAdLoadAndShow
            }
            print("在 didSet 中, value = \(isAdLoadAndShow), oldValue = \(oldValue)")
        }
    }
    
    
    // 是否展示某些按钮
    public var isOnlyLoad: Bool = false {
      
        didSet {
            if !isOnlyLoad {
                btnLoad.isHidden = !isOnlyLoad
                btnShow.isHidden = !isOnlyLoad
            }
            print("在 didSet 中, value = \(isOnlyLoad), oldValue = \(oldValue)")
        }
    }
    
    override func viewDidLoad() {
        super.viewDidLoad()
        self.versionAdaptation()
        self.setupUI()
        loadAdWithState(._Normal)
    }
    
    // MARK: - setupUI
    func setupUI() {
        self.view.backgroundColor = .white
        self.view.addSubview(self.btnLoad)
        self.view.addSubview(self.btnShow)
        self.view.addSubview(self.btnLoadAndShow)
        self.view.addSubview(self.labNotify)
        self.view.addSubview(self.textV)
    }
    
    // MARK: - event
    // 加载广告
    @objc func handleClickLoadBtn() {
        self.clearText()
        loadAdWithState(._Normal)
    }
    // 显示广告
    @objc func handleClickShowBtn() {
        
    }
    // 加载并显示广告
    @objc func handleClickLoadAndShowBtn() {
        self.clearText()
        loadAdWithState(._Normal)
    }
    // 释放
    func deallocAd() {
        
    }
    // 加载状态
    func loadAdWithState(_ state: AdState) {
        self.labNotify.text = ""
        switch (state) {
        case ._Normal:
            self.labNotify.text = "未加载广告"
        case ._Loading:
            self.labNotify.text = "广告加载中..."
        case ._LoadSucceed:
            self.labNotify.text = "广告加载成功"
        case ._LoadFailed:
            self.labNotify.text = "广告加载失败"
        }
    }
    
    func showProcessWithText(_ text: String) {
        tempText = (tempText as String) + "\r\n\(text)"
        self.textV.text = tempText as String;
    }
    
    func clearText() {
        tempText = ""
        self.textV.text = tempText as String
    }
    
    func showErrorWithDescription(_ description: NSDictionary?) {
        
        if description == nil {
            return
        }
    
        let highlightText = self.getSianKey(withDic: description as? [String : NSError])
        tempText = tempText.appending(String(", 失败原因如下:\r\n\(highlightText)"))
        self.textV.text = tempText as String
    }
    
    // MARK: - prinvate
    /// 适配iOS
    func versionAdaptation() {
        if #available(iOS 15, *) {
            let appearance: UINavigationBarAppearance = UINavigationBarAppearance()
            appearance.backgroundEffect = UIBlurEffect(style: .regular)
            self.navigationController?.navigationBar.scrollEdgeAppearance = appearance
            self.navigationController?.navigationBar.standardAppearance = appearance
        }
    }
    
    // 拼接错误内容返回字符串
    func getSianKey(withDic dic: [String : NSError]?) -> String  {
        //按字典排序
        let arr = dic?.keys
        guard arr != nil else {
            return ""
        }
        
        let tempArr = arr?.sorted(by: { a1, a2 in
            let result = a1.compare(a2)
            return result == .orderedDescending
        })
        
        guard tempArr != nil else {
            print(tempArr ?? "为空")
            return ""
        }
        
        //拼接字符串
        var strArray: [String] = []
        for(_ , value) in tempArr!.enumerated() {

            var appendStr = ""
            appendStr = value
            let par: NSError? = dic?[value] as? NSError ?? nil
            
            if dic?[value] != nil {
                appendStr = appendStr + (par?.userInfo.description ?? "")
                let code = par?.code;
                var desc = par?.userInfo[NSLocalizedDescriptionKey];
                
                if desc == nil {
                    desc = par?.userInfo["desc"]
                }
                
                let errorInfo = "\(value): code: \(String(describing: code ?? 0)) 错误详情:\(String(describing: desc))"
                strArray.append(errorInfo)
           }
        }
        let str = strArray.joined(separator: "\r\n")
        return str
    }
}
